### PR TITLE
feat: 🎸 Added RestoreAcrs authproc filter, modify ACRs when MFA

### PR DESCRIPTION
**Pull request by dBucik**

### CODE REVIEW

**Comment by dBucik** [comment]
Unfortunately, the authproc filter is heavily tied to the actions done in the WAYF. Do you have an idea, how could it be written in a better, more as a one-piece way? @melanger @vyskocilpavel

**Comment by melanger** [comment]
> Unfortunately, the authproc filter is heavily tied to the actions done in the WAYF. Do you have an idea, how could it be written in a better, more as a one-piece way? @melanger @vyskocilpavel

I guess you are using `disco.php` as an "external" discovery URL. This is probably one of the best ways to do it. You can also look at the `extDiscoveryStorage` option, but I think it would work basically the same.

You can refactor the code a bit - keep the original Disco class and extend it (subclass) with the MFA authncontext handling. This would allow somebody else with a different discovery service to add the MFA handling without switching to this discovery service.

There might be a proxy without discovery service. I guess for this case you can make a fake "discovery service" which does not do the discovery, it only modifies authncontext and forwards the request.

Generally, I would keep the "discovery service with MFA capabilities" as it is, and add a note to documentation that it is not possible to use the MFA forwarding function with another WAYF or without WAYF at the moment.

File: lib/Auth/Process/RestoreAcrs.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        parent::__construct($config, $reserved);
+    }
+
+    public function process(&$request)
+    {
+        $this->restoreAcrs($request);
+    }
+
+    public static function storeAcrs(array &$state, array $acrsToAdd)
```
Comment (melanger) [inline lib/Auth/Process/RestoreAcrs.php:33]:
```suggestion
    public static function addAndStoreAcrs(array &$state, array $acrsToAdd)
```
Might be a more suitable name.

File: lib/Auth/Process/RestoreAcrs.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+use SimpleSAML\Logger;
+use SimpleSAML\Module\perun\Disco;
+
+/**
+ * Auth proc filter, which should be used if Proxy provides MFA, and wants to first pass the request to the upstream
+ * IdP. The Perun Disco page, tries to add new ACRs to the request, if MFA has been requested, to prevent the IdP from
+ * failing. As a result, we then get modified requested ACRs, which should be restored to the previous (original) state
+ * using this authproc filter. It should be run on one of the first places of the IdP authproc chain.
+ */
+class RestoreAcrs extends ProcessingFilter
```
Comment (melanger) [inline lib/Auth/Process/RestoreAcrs.php:17]:
```suggestion
class multifactorAcrs extends ProcessingFilter
```
It's just that `RestoreAcrs::storeAcrs` feels weird.

File: lib/Auth/Process/RestoreAcrs.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            $state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF] = $new;
+        }
+    }
+
+    private function restoreAcrs(&$request)
+    {
+        if (!empty($request[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL])) {
+            unset($request[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF]);
+            $handle = &$request[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF];
+            $handle = $request[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL];
```
Comment (melanger) [inline lib/Auth/Process/RestoreAcrs.php:74]: I don't think this is right. You override `$handle` here without using the first value. Also, the `&` in `&$request` should not be there, it is in the function signature.

File: lib/Disco.php [hunk 4]
```diff
@@ -970,4 +977,10 @@ private function fillSpName($t)
             //OK, we will just display the disco
         }
     }
+
+    private function prepareAcrsForMfa(array &$state)
+    {
+        $contextsToAdd = $this->wayfConfiguration->getArray(self::ADD_AUTHN_CONTEXT_CLASSES_FOR_MFA, []);
+        Module\perun\Auth\Process\RestoreAcrs::storeAcrs($state, $contextsToAdd);
```
Comment (melanger) [inline lib/Disco.php:984]:
```suggestion
        RestoreAcrs::storeAcrs($state, $contextsToAdd);
```
+ add `use SimpleSAML\Module\perun\Auth\Process\RestoreAcrs;` at the top

**Comment by dBucik** [comment]
Can be reviewed and merged if OK. Code has been tested on ELIXIRs Proxy, seems to work just fine.

**Review by melanger — APPROVED** [review]
@dBucik I don't have a better idea where to put the code other than the discovery service. We could make a PR for SSP to add a new hook which runs soon enough, but since we use this disco on all proxies, it brings no benefits.

I would just add this auth proc filter to documentation and describe what needs to be done (that this filter only works with the disco + if the disco is documented, add a note there that this filter should be used).

**Pull request merged**